File: app/src/main/java/com/example/marosu/secretchat/base/BaseViewModel.java
package com.example.marosu.secretchat.base;

import android.arch.lifecycle.ViewModel;

/**
 * Created by Marius-Andrei Rosu on 8/7/2017.
 */
public final class BaseViewModel<V extends BaseContract.View, P extends BaseContract.Presenter<V>> extends ViewModel {
    private P presenter;

    public P getPresenter() {
        return this.presenter;
    }

    void setPresenter(P presenter) {
        if (this.presenter == null) {
            this.presenter = presenter;
        }
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        presenter.onPresenterDestroy();
        presenter = null;
    }
}

File: app/src/main/java/com/example/marosu/secretchat/base/BaseViewModel.kt
package com.example.marosu.secretchat.base

import android.arch.lifecycle.ViewModel

/**
 * Created by Marius-Andrei Rosu on 8/29/2017.
 */
class BaseViewModel<V : BaseContract.View, P : BaseContract.Presenter<V>> : ViewModel() {
    var presenter: P? = null

    override fun onCleared() {
        super.onCleared()
        presenter?.onPresenterDestroy()
        presenter = null
    }
}
